fprintd, libfprint : update to 1.90.1

Elyhaka opened the pull request:
###### Motivation for this change

Updates `libfprint` and `fprintd`. It also adds `libpam-wrapper` : a library made to test pam modules. This library is used by upstream to test the `fprintd` build.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @abbradar

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -47,13 +45,28 @@ stdenv.mkDerivation rec {
     libfprint
   ];
 
+  preConfigure = "./autogen.sh";
```
```suggestion
  preConfigure = ''
    NOCONFIGURE=1 ./autogen.sh
  '';
```

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,40 +1,38 @@
-  src = fetchurl {
-    url = "https://gitlab.freedesktop.org/libfprint/fprintd/uploads/9dec4b63d1f00e637070be1477ce63c0/fprintd-${version}.tar.xz";
-    sha256 = "182gcnwb6zjwmk0dn562rjmpbk7ac7dhipbfdhfic2sn1jzis49p";
+  src = fetchgit {
```
You can use `fetchFromGitLab`.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -47,13 +45,28 @@ stdenv.mkDerivation rec {
+  preConfigure = "./autogen.sh";
+
+  postPatch = ''
+    substituteInPlace tests/fprintd.py \
```
Maybe add test python to `checkInputs`

Elyhaka replied:
It did not work (test fails like there was no shebang substitutes), am I missing something ?

jtojnar reviewed (commented):
The MR was merged, release still not tagged though.

Elyhaka commented:
Hello @jtojnar : thank you for your review. Currently upstream does not compile, I believe this is the reason they did not tagged it yet.

Elyhaka commented:
We have two MR on upstream for releasing the 1.90.1 on both the library and the binary:
 - `libfprint` : https://gitlab.freedesktop.org/libfprint/libfprint/merge_requests/108
 - `fprintd` : https://gitlab.freedesktop.org/libfprint/fprintd/merge_requests/19

Elyhaka commented:
I've pushed an update with the latest content of the merge requests on GitLab.
I'm having trouble to add `libpam-wrapper` (A helper to test PAM modules added in latest upstream commits) into the python context (this C library is also building a Python library with it).
If somebody knows how to handle that specific case :).

Elyhaka commented:
1.90.1 version has been released on both projects 3 hours ago, I'm currently working on integrating it. I will submit this PR asap. Is it too late for the feature freeze of 20.03 ?

Elyhaka commented:
Finished to port `1.90.1` version :

 - `libfprint` : Version `1.90.1`
 - `libpam-wrapper` : New dependency for building/testing the software
 - `fprintd` : Version `1.90.1`

Current testing :
 - Worked well for enrolling and using it with on my Lenovo Thinkpad T495 (`06cb:00bd` device).

The `1.90.x` adds native support for a lot of devices that are already on the market, and I believe it could be good thing to have it in 20.03.

worldofpeace commented:
We can only backport bugfixes releases, I'm not sure this qualifies as one?

jtojnar commented:
Given that this is still a pre-release, I would move this to 20.09

Elyhaka commented:
@jtojnar : I don't see any mention that this one is deemed a pre-release ? They seems to have wanted to release a 1.90.1 version to support the devices while iterating toward a 2.0 version ([based on comment on this MR on `libfprint`](https://gitlab.freedesktop.org/libfprint/fprintd/merge_requests/19))

But let's keep it for 20.09 anyway ;)

Elyhaka commented:
Bonus question : should I add myself to the maintainers ?

worldofpeace commented:
> Bonus question : should I add myself to the maintainers ?

Totally

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/working-on-updating-libfprint-to-1-90/4958/3

LEXUGE commented:
@Elyhaka 
On my X1 Carbon 7th Gen (20R1), a patch is needed for it to enroll. I successfully resolved yesterday's problem by using [patch](https://aur.archlinux.org/cgit/aur.git/tree/disable-systemd-protection.patch?h=fprintd-libfprint2)

jtojnar commented:
@LEXUGE Is there an upstream bug report/merge request for that?

LEXUGE commented:
> @LEXUGE Is there an upstream bug report/merge request for that?

Not really but it should be a bug

jtojnar commented:
Could you open it then please? Applying random patches for unreported bugs rarely leads to upstream fixes.

LEXUGE commented:
> Could you open it then please? Applying random patches for unreported bugs rarely leads to upstream fixes.

I will do it later.

LEXUGE commented:
Issue has been opened
https://gitlab.freedesktop.org/libfprint/fprintd/issues/50

Elyhaka commented:
@LEXUGE : While being an issue on upstream, the way I pass the path to the derivation seems to be broken as is. I'm not sure why it works on my side, probably because I have a state directory that is populated by many tries of many versions while working on the derivation.

Elyhaka commented:
I've checked the `meson` documentation and it seems that the `/` operator joins path together. I have the feeling that when two joined path starts with `/` it adds a `:`. **But I am not sure it has behavior at all**. I am really not used to work with this build system at all.

LEXUGE commented:
So has the issue of broken path solved?

Elyhaka commented:
No, not yet, I'd like a second opinion on this subject.

LEXUGE commented:
@Elyhaka I'm still not clear about you mean by broken path. Does that mean your packaging work on current build but will fail in the future because the way you package would be broken?
On my side it's working too.

Elyhaka commented:
@LEXUGE If I refer to the issue you opened, it seems that the storage path has the format `path1:path2` on your system. I'm not sure why it did this on your system to be honest.
Upstream is working on this support for version `2.0`.

LEXUGE commented:
> @LEXUGE If I refer to the issue you opened, it seems that the storage path has the format `path1:path2` on your system. I'm not sure why it did this on your system to be honest.
> Upstream is working on this support for version `2.0`.

emmmm, but on other systems like Arch, there are also this issue as well.
And after the dirty patch, it worked.

Elyhaka commented:
Updated this PR:

 - Rebased on latest master
 - Cleaned up new derivation `libpam-wrapper` using the upstream version (and not the fork made by `fprintd` developers)

The issue @LEXUGE mentionned has been fixed [by this commit](https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/fd733e55be8f407bf849b557d03dc2fa41e66907). I don't know if I should add it as a patch over the 1.90.1 version or simply wait for the 2.0 release.

matejc commented:
This is marked "Works for me" by me, on thinkpad t490s, I had to upgrade my firmware for the fingerprint also (https://github.com/fwupd/fwupd/wiki/LVFS-Testing-remote).. but otherwise... works great...

LEXUGE commented:
> Updated this PR:
> 
>     * Rebased on latest master
> 
>     * Cleaned up new derivation `libpam-wrapper` using the upstream version (and not the fork made by `fprintd` developers)
> 
> 
> The issue @LEXUGE mentionned has been fixed [by this commit](https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/fd733e55be8f407bf849b557d03dc2fa41e66907). I don't know if I should add it as a patch over the 1.90.1 version or simply wait for the 2.0 release.

Yes, I filed an issue and it has been fixed.

Elyhaka commented:
> This is marked "Works for me" by me, on thinkpad t490s, I had to upgrade my firmware for the fingerprint also (https://github.com/fwupd/fwupd/wiki/LVFS-Testing-remote).. but otherwise... works great...

Yes, for some synaptics peripherals, the firmware is not yet out of LVFS testing.

Elyhaka commented:
ArchLinux switched to 1.90.1 on `fprintd` and `libfprint`. I'm using it daily for the last months, maybe we could check if that makes sense for `unstable` on our side ?

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,35 @@
+  pname = "libpam-wrapper";
+
+  src = fetchFromGitLab {
+    domain = "gitlab.com";
```
This is the default.

jtojnar replied:
Actually, I see no mention of GitLab on the website: https://cwrap.org/download.html#pam_wrapper

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,35 @@
+    domain = "gitlab.com";
+    owner = "cwrap";
+    repo = "pam_wrapper";
+    rev = "ab2d90c6def33888db86755e23f3c12d417e187f";
```
Using tags is easier to update:

```suggestion
    rev = "pam_wrapper-${version}";
```

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,35 @@
+  buildInputs = [ openpam ];
+
+  cmakeFlags = [
+    "-DMAN_INSTALL_DIR=$out/share/man/man1"
```
CMake does not interpolate variables in `-D` flags so it would have to be:

```suggestion
    "-DMAN_INSTALL_DIR=${placeholder "out"}/share/man/man1"
```

But the project uses GNUInstallDirs

https://git.samba.org/?p=pam_wrapper.git;a=blob;f=CMakeLists.txt;h=b453ec361b19ac17dc6f1de6266999d4d4a8cdbe;hb=ab2d90c6def33888db86755e23f3c12d417e187f#l40

so the option would be `CMAKE_INSTALL_MANDIR`, which we already set:

https://github.com/NixOS/nixpkgs/blob/85d6f3bcd9cbcc52c4a307d2ef5116dab4b41641/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L78

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,35 @@
+  meta = with stdenv.lib; {
+    description = "Wrapper for testing pam modules.";
+    homepage = "https://cwrap.org/pam_wrapper.html";
+    license = licenses.gpl3;
```
```suggestion
    license = licenses.gpl3Plus;
```


https://git.samba.org/?p=pam_wrapper.git;a=blob;f=src/libpamtest.c;h=4474736d688bd0fe9652cac1ba909fbdf05f502e;hb=ab2d90c6def33888db86755e23f3c12d417e187f#l7

Elyhaka commented:
@jtojnar Thank you for the review! I've made the asked changes.

jtojnar commented on pkgs/development/libraries/libfprint/default.nix:
```diff
@@ -1,41 +1,30 @@
+    domain = "gitlab.freedesktop.org";
+    owner = "libfprint";
+    repo = "libfprint";
+    rev = "66c9e4a829a06a25d8b6160cdfbad1d47ef5b81a";
```
similarly, this should be a tag, or the version should be `unstable-yyyy-mm-dd`

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,29 @@
+
+stdenv.mkDerivation rec {
+  version = "1.1.3";
+  pname = "libpam-wrapper";
```
pname is by convention first.

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,29 @@
+  buildInputs = [ openpam ];
+
+  meta = with stdenv.lib; {
+    description = "Wrapper for testing pam modules.";
```
```suggestion
    description = "Wrapper for testing PAM modules";
```

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,29 @@
+
+  meta = with stdenv.lib; {
+    description = "Wrapper for testing pam modules.";
+    homepage = https://github.com/jhrozek/pam_wrapper;
```
```suggestion
    homepage = "https://github.com/jhrozek/pam_wrapper";
```

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -46,23 +54,36 @@ stdenv.mkDerivation rec {
+      "'$out/share/polkit-1/actions'"
+
+    substituteInPlace meson.build --replace \
+      "dbus_dep.get_pkgconfig_variable('interfaces_dir')" \
```
See `man pkg-config`:

```
       PKG_CONFIG_$PACKAGE_$VARIABLE
              Overrides  the  variable  VARIABLE in the package PACKAGE. The environment variable should have the package name and package variable upper cased with non-alphanumeric characters converted to
              underscores. For example, setting PKG_CONFIG_GLADEUI_2_0_CATALOGDIR will override the variable "catalogdir" in the "gladeui-2.0" package.
```

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,50 @@
+    pygobject3
+    pycairo
+    libpam-wrapper
+    gtk-doc
```
These are not python deps.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,50 @@
-  version = "0.9.0";
+  pname = "fprintd";
+  version = "1.90.1";
 
```
gtk-doc will not be persisted without devdoc output:

```suggestion

  outputs = [ "out" "devdoc" ];

```

Elyhaka commented:
@jtojnar Thank you again for your review, I've tested and pushed the update. I've also added the patch that upstream has made to address @LEXUGE's issue.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,66 @@
+    pythonPath
+    gtk-doc
+    libxslt
+    docbook_xsl
```
This is alias, the proper name is `docbook-xsl-nons`.

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,29 @@
+    maintainers = [ maintainers.elyhaka ];
+    platforms = platforms.linux;
+  };
+}
```
Missing `\n`.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,66 @@
-    gtk-doc # Drop with above patch
+    meson
+    ninja
+    intltool
```
```suggestion
    gettext
```
https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/53d80c14742b8dd387aa449360e2b68bc0291e64

Elyhaka replied:
I've tried to remove `intltool` to put `gettext` only, but it fails with the following error:

```
meson.build:84:0: ERROR: Program(s) ['pod2man'] not found or not executable
```

jtojnar replied:
then you need to add `perl # for pod2man` as well.

Elyhaka replied:
That was it, thank you.

jtojnar commented:
Thinkpad mention should probably be removed from nixos/modules/services/security/fprintd.nix

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,68 @@
+    pythonPath
+    gtk-doc
+    libxslt
+    docbook-xsl-nons
```
Fix validation warning:

```
I/O error : Attempt to load network entity http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd
../fprintd-docs.xml:5: warning: failed to load external entity "http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd"
]>
  ^


```
```suggestion
    docbook-xsl-nons
    docbook_xml_dtd_412
```

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,70 @@
+    gtk-doc
+    libxslt
+    docbook-xsl-nons
+    docbook_xml_dtd_412
```
Maybe also `dbus # for dtd`:

```
/build/source/build/src/net.reactivated.Fprint.Manager.xml:5: warning: failed to load external entity "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd"
]>
  ^
```

But not sure if this one will be picked up automatically.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -46,23 +74,27 @@ stdenv.mkDerivation rec {
+    "-Dpam_modules_dir=${placeholder "out"}/lib/security"
+    "-Dsysconfdir=${placeholder "out"}/etc"
+    "-Ddbus_service_dir=${placeholder "out"}/share/dbus-1/system-services"
+    "-Dpolicydir=${placeholder "out"}/share/polkit-1/actions"
```
`WARNING: Unknown options: "policydir"
`

jtojnar commented on pkgs/development/libraries/libfprint/default.nix:
```diff
@@ -44,17 +33,15 @@ stdenv.mkDerivation rec {
+  ];
 
   mesonFlags = [
     "-Dudev_rules_dir=${placeholder "out"}/lib/udev/rules.d"
```
WARNING: Unknown options: "x11-examples"

jtojnar replied:
substituteStream(): WARNING: pattern '/bin/echo' doesn't match anything in file 'libfprint/meson.build'

jtojnar commented on pkgs/development/libraries/libfprint/default.nix:
```diff
@@ -1,41 +1,30 @@
-  }.${pname};
+  pname = "libfprint";
+  version = "1.90.1";
+
```
> Removing /nix/store/n110wj82kgnmx1ang2zpadp1w9bhrr33-libfprint-1.90.1/share/gtk-doc


```suggestion

  outputs = [ "out" "devdoc" ];

```

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,70 @@
+    sha256 = "0mbzk263x7f58i9cxhs44mrngs7zw5wkm62j5r6xlcidhmfn03cg";
   };
 
+  outputs = [ "out" "devdoc" ];
```
This should go between version and src, see https://discourse.nixos.org/t/document-attribute-ordering-in-package-expressions/4887.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,70 @@
+    perl
+    gettext
+    libpam-wrapper
+    pythonPath
```
Inputs cannot contain lists and it will be an hard error with structured attrs.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,70 @@
+    dbus-python
+    pygobject3
+    pycairo
+    libpam-wrapper
```
This will super-likely lead to a python mismatch. Add it to as `pypamtest` to `python-packages.nix` and override the python there, see for example:

https://github.com/NixOS/nixpkgs/blob/28c2c0156da98dbe553490f904da92ed436df134/pkgs/top-level/python-packages.nix#L3059-L3075

Elyhaka replied:
Noted, should I simply do an override like:

```nix
  pypamtest = pkgs.libpam-wrapper.override {
    enablePython = true;
    inherit python;
  };
```

Or should I split the output with a `py` one and add the whole `pipe` like `libselinux` ? I may have to patch/substitute some CMake files on `libpam-wrapper` because they always prefix the output path for python with the `CMAKE_INSTALL_PREFIX`.

jtojnar replied:
If this is only used for build/tests, splitting might be nice but not be necessary.

Elyhaka replied:
`libpam-wrapper` is made for testing only. I will try to split it later if that's not a blocker :)

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,70 @@
     })
   ];
 
+  pythonPath = with python37Packages; [
```
Is Python 3.7 specifically needed or any Python 3 version will work? Either use `python3.pkgs` or add a comment.

Elyhaka replied:
It should work with all python3 version.

Elyhaka commented:
@jtojnar Thank you very much for your time and your extensive review, let me know if there is other parts I can improve.

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,34 @@
+, python ? null
+}:
+
+with stdenv.lib;
```
`with` statement makes it harder to reason about code. Prefer using qualified names, `let` with `inherit` or narrow the scope of `with` as much as possible.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,43 +1,67 @@
+    dbus
+    docbook-xsl-nons
+    docbook_xml_dtd_412
+  ] ++ (with python3.pkgs; [
```
Ideally, these would be in `checkInputs` but that would require `doCheck = true`. Maybe add a TODO comment?

Elyhaka replied:
I added the TODO, I've tried to move it, but it breaks things. Out of curiosity, what would be the requirement to accomplish such modification ? I have the feeling that I should disable the tests from the buildPhase, and write a custom `checkPhase` right ?

jtojnar replied:
Ideally, you would figure out how to fix the tests – often it is just a matter of running them inside dbus session:

https://github.com/NixOS/nixpkgs/blob/4b4e2b390ae53e1213081504b201a8e0e2e566f7/pkgs/applications/networking/calls/default.nix#L69-L76

(not all these lines necessary)

but `checkPhase = ":";` (`:` = shell null command) with a `TODO` comment would work too.

Elyhaka replied:
With the `checkPhase = ":"` it still fails with:

```
ERROR: Dependency "pam_wrapper" not found, tried pkgconfig
```

I will play with this next week-end if that's not a blocker :).

Elyhaka commented:
I've redone a round of tests (enrolling, login with different managers like GDM etc...) it works correctly on my side.

jtojnar commented:
```patch
diff --git a/pkgs/tools/security/fprintd/default.nix b/pkgs/tools/security/fprintd/default.nix
index f589ffe8c76..0f5b2ac2244 100644
--- a/pkgs/tools/security/fprintd/default.nix
+++ b/pkgs/tools/security/fprintd/default.nix
@@ -20,7 +20,6 @@
 , systemd
 , libfprint
 , python3
-, pypamtest
 }:
 
 stdenv.mkDerivation rec {
@@ -43,6 +42,20 @@ stdenv.mkDerivation rec {
       url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/merge_requests/50.patch";
       sha256 = "1wqkk7bqig0fhjbq1487kxkh7yc9356qjsac1x27pzgakbayxpyy";
     })
+    # Fixes issue with ":" when there is multiple paths (might be the case on NixOS)
+    # https://gitlab.freedesktop.org/libfprint/fprintd/-/merge_requests/50
+    (fetchpatch {
+      url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/d7fec03f24d10f88d34581c72f0eef201f5eafac.patch";
+      sha256 = "QNN05WF4YZ0XiTwm5NkfqZDuQpyXlnrh+RJF9SNsCDk=";
+    })
+    (fetchpatch {
+      url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/4f3589c0dc0feb67c45cdae7e692d70fe493f3d8.patch";
+      sha256 = "2kYo8uqMja3r9SroKB5cR777MJ/BGig4PB1qwLUQ01U=";
+    })
+    (fetchpatch {
+      url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/f401f399a85dbeb2de165b9b9162eb552ab6eea7.patch";
+      sha256 = "Pga+/QEkln8DOwGZfKM1r2urJX4Y3X0bozWWxKZ5ia0=";
+    })
   ];
 
   nativeBuildInputs = [
@@ -91,7 +104,10 @@ stdenv.mkDerivation rec {
   # Ideally, we should find a way to execute the test here
   # instead of the buildPhase.
   doCheck = true;
-  checkPhase = ":";
+
+  postPatch = ''
+    patchShebangs po/check-translations.sh
+  '';
 
   meta = with stdenv.lib; {
     homepage = "https://fprint.freedesktop.org/";

```

Elyhaka commented:
@jtojnar IIRC, GCC checks the `LIBRARY_PATH` folder. I don't know how acceptable is this solution, but adding:

```
LIBRARY_PATH = stdenv.lib.makeLibraryPath [ libpam-wrapper ];
```

Fixes this issue.

jtojnar commented:
I think it is okay for tests in `preCheck`, it should not affect things much. But I wonder why they do not use `compiler.find_library()`.

Elyhaka commented:
I'm now fighting with a `python` program not loaded correctly from meson:

```
tests/meson.build:50:0: ERROR: Program(s) ['unittest_inspector.py'] not found or not executable
```

I'm not sure how to fix this one.

jtojnar commented:
Missed two commits from https://gitlab.freedesktop.org/libfprint/fprintd/-/merge_requests/40, now picking it completely:

```patch
diff --git a/pkgs/tools/security/fprintd/default.nix b/pkgs/tools/security/fprintd/default.nix
index 3f25e813279..f5bb97a65a5 100644
--- a/pkgs/tools/security/fprintd/default.nix
+++ b/pkgs/tools/security/fprintd/default.nix
@@ -42,6 +42,20 @@ stdenv.mkDerivation rec {
       url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/merge_requests/50.patch";
       sha256 = "1wqkk7bqig0fhjbq1487kxkh7yc9356qjsac1x27pzgakbayxpyy";
     })
+    # Fixes issue with ":" when there is multiple paths (might be the case on NixOS)
+    # https://gitlab.freedesktop.org/libfprint/fprintd/-/merge_requests/50
+    (fetchpatch {
+      url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/d7fec03f24d10f88d34581c72f0eef201f5eafac.patch";
+      sha256 = "QNN05WF4YZ0XiTwm5NkfqZDuQpyXlnrh+RJF9SNsCDk=";
+    })
+    (fetchpatch {
+      url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/merge_requests/40.patch";
+      sha256 = "43uPihK6HhygHw1Qplwci80Wseq/S77VUp+OdEECHmM=";
+    })
+    (fetchpatch {
+      url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/f401f399a85dbeb2de165b9b9162eb552ab6eea7.patch";
+      sha256 = "Pga+/QEkln8DOwGZfKM1r2urJX4Y3X0bozWWxKZ5ia0=";
+    })
   ];
 
   nativeBuildInputs = [
@@ -75,10 +89,6 @@ stdenv.mkDerivation rec {
     pypamtest
   ];
 
-  PKG_CONFIG_DBUS_1_INTERFACES_DIR = "${placeholder "out"}/share/dbus-1/interfaces";
-  PKG_CONFIG_POLKIT_GOBJECT_1_POLICYDIR = "${placeholder "out"}/share/polkit-1/actions";
-  PKG_CONFIG_DBUS_1_DATADIR = "${placeholder "out"}/share";
-
   mesonFlags = [
     "-Dgtk_doc=true"
     "-Dpam_modules_dir=${placeholder "out"}/lib/security"
@@ -87,10 +97,18 @@ stdenv.mkDerivation rec {
     "-Dsystemd_system_unit_dir=${placeholder "out"}/lib/systemd/system"
   ];
 
-  # Ideally, we should find a way to execute the test here
-  # instead of the buildPhase.
+  PKG_CONFIG_DBUS_1_INTERFACES_DIR = "${placeholder "out"}/share/dbus-1/interfaces";
+  PKG_CONFIG_POLKIT_GOBJECT_1_POLICYDIR = "${placeholder "out"}/share/polkit-1/actions";
+  PKG_CONFIG_DBUS_1_DATADIR = "${placeholder "out"}/share";
+
+  # FIXME: Ugly hack for tests to find libpam_wrapper.so
+  LIBRARY_PATH = stdenv.lib.makeLibraryPath [ python3.pkgs.pypamtest ];
+
   doCheck = true;
-  checkPhase = ":";
+
+  postPatch = ''
+    patchShebangs po/check-translations.sh
+  '';
 
   meta = with stdenv.lib; {
     homepage = "https://fprint.freedesktop.org/";

```

But it still fails and the errors are even less helpful:

```
21/26 fprintd:PAM+test_pam_fprintd+TestPamFprintd / TestPamFprintd.test_pam_fprintd_auth                                 FAIL           0.97s (exit status 1)

--- command ---
08:42:36 LD_PRELOAD='/nix/store/7a2q4pykjq9c0jjmdf2v96wjpsvzy8i1-libpam-wrapper-1.1.3/lib/libpam_wrapper.so' TOPSRCDIR='/build/source' TOPBUILDDIR='/build/source/build' G_SLICE='always-malloc' MALLOC_CHECK_='2' G_DEBUG='fatal-warnings' PAM_WRAPPER_SERVICE_DIR='/build/source/build/tests/pam/services' PAM_WRAPPER_DEBUGLEVEL='2' PAM_WRAPPER='1' MALLOC_PERTURB_='55' /nix/store/2dcsn57cgaxs92ha5swihrab0g3l2h6g-python3-3.7.7/bin/python3 /build/source/tests/pam/test_pam_fprintd.py TestPamFprintd.test_pam_fprintd_auth
--- stdout ---
Using template from /build/source/tests/dbusmock/fprintd.py
test_pam_fprintd_auth (__main__.TestPamFprintd) ... ERROR

======================================================================
ERROR: test_pam_fprintd_auth (__main__.TestPamFprintd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/source/tests/pam/test_pam_fprintd.py", line 99, in test_pam_fprintd_auth
    res = pypamtest.run_pamtest("toto", "fprintd-pam-test", [tc], [ 'unused' ])
pypamtest.PamTestError: [Errno 1] pam_start failed()

----------------------------------------------------------------------
Ran 1 test in 0.649s

FAILED (errors=1)
--- stderr ---
PWRAP_DEBUG[<unknown> (965)] - pwrap_init: Initialize pam_wrapper
PWRAP_DEBUG[<unknown> (965)] - pwrap_init: Copy /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0 to /tmp/pam.g/lib/libpam.so.0
PWRAP_DEBUG[<unknown> (965)] - copy_confdir: Copy config files from /build/source/build/tests/pam/services to /tmp/pam.g
PWRAP_DEBUG[<unknown> (965)] - pwrap_init: Successfully initialized pam_wrapper
PWRAP_DEBUG[<unknown> (967)] - pwrap_init: Initialize pam_wrapper
PWRAP_DEBUG[<unknown> (967)] - pwrap_init: Copy /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0 to /tmp/pam.i/lib/libpam.so.0
PWRAP_DEBUG[<unknown> (967)] - copy_confdir: Copy config files from /build/source/build/tests/pam/services to /tmp/pam.i
PWRAP_DEBUG[<unknown> (967)] - pwrap_init: Successfully initialized pam_wrapper
PWRAP_DEBUG[<unknown> (970)] - pwrap_init: Initialize pam_wrapper
PWRAP_DEBUG[<unknown> (970)] - pwrap_init: Copy /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0 to /tmp/pam.l/lib/libpam.so.0
PWRAP_DEBUG[<unknown> (970)] - copy_confdir: Copy config files from /build/source/build/tests/pam/services to /tmp/pam.l
PWRAP_DEBUG[<unknown> (970)] - pwrap_init: Successfully initialized pam_wrapper
PWRAP_DEBUG[<unknown> (965)] - pwrap_load_lib_handle: Opened /tmp/pam.g/lib/libpam.so.0

-------
```

Elyhaka commented:
I haven't been able to fix the tests, the reason for the `pam_start` failure is not logged anywhere, I could not manage to find it. I also tried to change libpam-wrapper version to see if the behavior changed (tried the version on Debian and ArchLinux) without any luck.

jtojnar commented:
Tried running it with

```
    substituteInPlace tests/pam/meson.build --replace PAM_WRAPPER_DEBUGLEVEL=2 PAM_WRAPPER_DEBUGLEVEL=3
```

in `postPatch` does not lead to much more info:

```
PWRAP_DEBUG[<unknown> (987)] - pwrap_init: Initialize pam_wrapper
PWRAP_TRACE[<unknown> (987)] - pwrap_init: pam_wrapper config dir: /tmp/pam.7
PWRAP_TRACE[<unknown> (987)] - pwrap_init: PAM path: /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so
PWRAP_TRACE[<unknown> (987)] - pwrap_init: PAM library: libpam.so.2.0.0
PWRAP_TRACE[<unknown> (987)] - pwrap_init: Reconstructed PAM path: /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0
PWRAP_DEBUG[<unknown> (987)] - pwrap_init: Copy /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0 to /tmp/pam.7/lib/libpam.so.0
PWRAP_TRACE[<unknown> (987)] - pwrap_init: Using libpam path: /tmp/pam.7/lib/libpam.so.0
PWRAP_DEBUG[<unknown> (987)] - copy_confdir: Copy config files from /build/source/build/tests/pam/services to /tmp/pam.7
PWRAP_TRACE[<unknown> (987)] - copy_ftw: Copying /build/source/build/tests/pam/services/fprintd-pam-test
PWRAP_DEBUG[<unknown> (987)] - pwrap_init: Successfully initialized pam_wrapper
PWRAP_DEBUG[<unknown> (1086)] - pwrap_init: Initialize pam_wrapper
PWRAP_TRACE[<unknown> (1086)] - pwrap_init: pam_wrapper config dir: /tmp/pam.T
PWRAP_TRACE[<unknown> (1086)] - pwrap_init: PAM path: /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so
PWRAP_TRACE[<unknown> (1086)] - pwrap_init: PAM library: libpam.so.2.0.0
PWRAP_TRACE[<unknown> (1086)] - pwrap_init: Reconstructed PAM path: /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0
PWRAP_DEBUG[<unknown> (1086)] - pwrap_init: Copy /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0 to /tmp/pam.T/lib/libpam.so.0
PWRAP_TRACE[<unknown> (1086)] - pwrap_init: Using libpam path: /tmp/pam.T/lib/libpam.so.0
PWRAP_DEBUG[<unknown> (1086)] - copy_confdir: Copy config files from /build/source/build/tests/pam/services to /tmp/pam.T
PWRAP_TRACE[<unknown> (1086)] - copy_ftw: Copying /build/source/build/tests/pam/services/fprintd-pam-test
PWRAP_DEBUG[<unknown> (1086)] - pwrap_init: Successfully initialized pam_wrapper
PWRAP_DEBUG[<unknown> (1149)] - pwrap_init: Initialize pam_wrapper
PWRAP_TRACE[<unknown> (1149)] - pwrap_init: pam_wrapper config dir: /tmp/pam.M
PWRAP_TRACE[<unknown> (1149)] - pwrap_init: PAM path: /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so
PWRAP_TRACE[<unknown> (1149)] - pwrap_init: PAM library: libpam.so.2.0.0
PWRAP_TRACE[<unknown> (1149)] - pwrap_init: Reconstructed PAM path: /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0
PWRAP_DEBUG[<unknown> (1149)] - pwrap_init: Copy /nix/store/cjwsxn95qzjbhxdilbb3g87xqdx66vvq-openpam-20170430/lib/libpam.so.2.0.0 to /tmp/pam.M/lib/libpam.so.0
PWRAP_TRACE[<unknown> (1149)] - pwrap_init: Using libpam path: /tmp/pam.M/lib/libpam.so.0
PWRAP_DEBUG[<unknown> (1149)] - copy_confdir: Copy config files from /build/source/build/tests/pam/services to /tmp/pam.M
PWRAP_TRACE[<unknown> (1149)] - copy_ftw: Copying /build/source/build/tests/pam/services/fprintd-pam-test
PWRAP_DEBUG[<unknown> (1149)] - pwrap_init: Successfully initialized pam_wrapper
PWRAP_TRACE[<unknown> (987)] - pwrap_pam_start: pam_start service=fprintd-pam-test, user=toto
PWRAP_DEBUG[<unknown> (987)] - pwrap_load_lib_handle: Opened /tmp/pam.7/lib/libpam.so.0

PWRAP_TRACE[<unknown> (987)] - pwrap_pam_start: pam_start rc=4
PWRAP_TRACE[<unknown> (987)] - pwrap_destructor: entering pwrap_destructor
PWRAP_TRACE[<unknown> (987)] - pwrap_destructor: destructor called for pam_wrapper dir /tmp/pam.7
PWRAP_TRACE[<unknown> (987)] - p_rmdirs_at: p_rmdirs_at removing /tmp/pam.7 at CWD

PWRAP_TRACE[<unknown> (987)] - p_rmdirs_at: p_rmdirs_at removing lib at fd=3
```

only that it must be failing somewhere in `libpam_pam_start`.


Replacing openpam by linux-pam seems to fix it, though.

Elyhaka commented:
I should switch the dependency from `openpam` to `linux-pam` in `libpam-wrapper` then ? I thought they were interchangeable seamlessly, it seems not.

Elyhaka commented:
@jtojnar I've pushed an update with the switch from `openpam` to `linux-pam`. `linux-pam` seems to be the de-facto standard for linux distributions, while `openpam` seems to be more related to the *BSD distributions, and is general provided as an alternative PAM implementation on Linux distributions.

jtojnar commented on pkgs/development/libraries/libpam-wrapper/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  nativeBuildInputs = [ cmake ] ++ stdenv.lib.optional enablePython [ python ];
+
+  buildInputs = [ linux-pam ];
```
Maybe add a comment that we need to use linux-pam otherwise it breaks fprintd.

jtojnar commented on pkgs/tools/security/fprintd/default.nix:
```diff
@@ -1,42 +1,68 @@
+      url = "https://gitlab.freedesktop.org/libfprint/fprintd/-/commit/d7fec03f24d10f88d34581c72f0eef201f5eafac.patch";
+      sha256 = "QNN05WF4YZ0XiTwm5NkfqZDuQpyXlnrh+RJF9SNsCDk=";
+    })
     (fetchpatch {
```
```suggestion

    # Fix locating libpam_wrapper for tests
    (fetchpatch {
```

Commit:
libpam-wrapper: init at 1.1.3

Commit:
libfprint: 1.0 -> 1.90.1

Commit:
fprintd: 0.9.0 -> 1.90.1

jtojnar reviewed (approved):
Looks good to me, thanks for all your hard work :superhero_woman:

Elyhaka commented:
Thanks a lot for your help, time and reviews!

Elyhaka commented:
Is anything still preventing this PR from being merged ? :)

jtojnar commented:
Should be fine, thanks again.

Elyhaka commented:
Thank you!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/t495-graphics-glitched/7950/6

veprbl commented:
See regression reported in #103739
